### ISSUE TITLE
Scalar functions throw exceptions even if wrapped in try()

### ISSUE THREAD

**Issue by kagamiori** [prestodb/presto#23150]
Some functions throw exceptions on invalid argument even if they are used inside try() that is supposed to suppress exceptions. Below is a list of these functions and the example queries:
| Functions | Example queries |
|-----------|------------------|
| contains | select try(contains(c0, c1)) from (values (array[array[1], array[null]], array[null])) t(c0, c1) |
| array_distinct | select try(array_distinct(c0)) from (values (array[array[1, null], array[1, null]])) t(c0) |
| array_remove | select try(array_remove(c0, c1)) from (values (array[array[1], array[null]], array[null])) t(c0, c1) |
| array_union | select try(array_union(c0, c1)) from (values (array[array[1], array[null]], array[array[null]])) t(c0, c1) |
| array_min/array_max | select try(array_max(c0)) from (values (array[array[1], array[null]], array[null])) t(c0, c1) |
| array_position | select try(array_position(c0, c1)) from (values (array[array[1], array[null]], array[null])) t(c0, c1) |
| transform_keys | select try(transform_keys(c0, (k, v) -> array[null])) from (values (map(array[1, 2], array[3, 4]))) t(c0) |

## Your Environment

* Presto version used: 0.288
* Storage (HDFS/S3/GCS..):
* Data source and connector used:
* Deployment (Cloud or On-prem):
* [Pastebin](https://pastebin.com/) link to the complete debug logs:

## Expected Behavior

I expect these queries to return NULL instead of throwing.

## Current Behavior

These queries throw exceptions.

## Steps to Reproduce

1.
2.
3.
4.

**Comment by kagamiori** [comment]
cc @mbasmanova @amitkdutta

**Comment by kagamiori** [comment]
I found that these functions do not always throw when input elements are arrays containing NULLs. It's likely because the function implementation compares the inner array sizes first and hence avoid comparing elements in the inner arrays. This, however, brings inconsistency as the implementation should be transparent to users.

| Functions | Example query |
|-----------|----------------|
| contains | select contains(c0, c1) from (values (array[array[1], array[1, null]], array[2])) t(c0, c1) |
| array_remove | select array_remove(c0, c1) from (values (array[array[1], array[1, null]], array[1])) t(c0, c1) |
| array_distinct | select array_distinct(c0) from (values (array[array[1], array[1, null]])) t(c0) |
| array_union | select array_union(c0, c1) from (values (array[array[1], array[1, null]], array[array[1], array[1]])) t(c0, c1) |
| array_position | select array_position(c0, c1) from (values (array[array[1], array[1, null]], array[2])) t(c0, c1) |

**Comment by kagamiori** [comment]
In the version 0.289, the following queries no longer throw exceptions on nested null elements. The rest functions stay the same.

| Functions | Example queries |
|-----------|------------------|
| array_distinct | select try(array_distinct(c0)) from (values (array[array[1, null], array[1, null]])) t(c0) |
| array_union | select try(array_union(c0, c1)) from (values (array[array[1], array[null]], array[array[null]])) t(c0, c1) |

**Comment by rschlussel** [comment]
array_distinct and array_union were changed as part of https://github.com/prestodb/presto/issues/22900

**Comment by kagamiori** [comment]
Hi @rschlussel, thank you for sharing the context! I wonder how the functions with "distinct" semantics are distinguished from functions with "equal" semantics? For example, should `contains`, `transform_keys`, and `map_subset` be considered to have the "distinct" semantic as well? 

Also, do you know why `TRY` is not able to suppress the error of comparison of arrays with null elements being unsupported, except for `array_sort`? I'd expect TRY to be able to suppress these errors since they are about invalid arguments to the functions.

**Comment by rschlussel** [comment]
maps do not support keys with indeterminate types (that's defined in the sql spec I think), so those need to throw. 

For the others. I'm not sure there's a clear definition, but in my head, I feel that if there's a corresponding operation on tables  the array function should match that (e.g. array_distinct should match select distinct, array_union should match select x union select y). Whereas contains feels more like "IN". I do think it's strange that we throw an exception for `contains(ARRAY[ARRAY[null]], ARRAY[null])`. I feel like the "right" thing to do here would be to return NULL, and it's just an implementation limitation that makes us throw. 


The reason try isn't catching these errors is because Presto is using the wrong error code when it's throwing the exceptions for those functions (we're throwing NOT_SUPPORTED errors https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayContains.java#L203, and should be throwing INVALID_FUNCTION_ARGUMENT errors, and the try code doesn't catch NOT_SUPPORTED errors https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/TryFunction.java#L166-L170). This is something that should be fixed in Presto.

**Comment by kagamiori** [comment]
@rschlussel I agree that `contains` should have the same semantic as `IN`. Though, I'm not sure whether `IN` and `contains` should have the "distinct" semantic now, because I noticed that a query `select a in (b)` is actually rewritten as `select a = b` in Presto, so there seems to be an assumption that `IN` (and hence `contains`) has the "equal" semantic.

Except `contains` that needs discussion, does the following make sense to us?
1. We will not change the behavior of the functions listed above in Presto, and will make Velox implementations for them match the Presto behavior.
2. In Velox, we'll let TRY suppress the exceptions thrown from these functions about the unsupported comparisons of arrays with null elements. (This is what Velox already does.)
3. In Presto, we'll expect a change to make TRY able to suppress exceptions about unsupported comparisons thrown from these functions. Then, the Presto and Prestissimo behavior will match.

cc @mbasmanova

**Comment by rschlussel** [comment]
yeah, i expect contains should have equals semantics (which would give null on null input, but I guess we throw instead of array[null]).  I found some history here: https://github.com/prestodb/presto/pull/10909.  

1. I think it might make sense to change Presto to return null on indeterminate input for contains and maybe some others (and changing to do looks pretty easy), but not sure, might need more discussion. Probably having velox copy presto semantics, as you suggested, is the simplest path forward.

2 and 3. Yes, I think that makes sense.

**Comment by kagamiori** [comment]
@rschlussel Yes, I agree that `contains` should returns NULL for array[null] to match `IN` behavior on the same input. Let me check which functions in Velox besides `contains` need to be fixed to match Presto. We can then share the plan of changing `contains` to get more opinions.

**Comment by kagamiori** [comment]
Three functions doesn't match Presto's current behavior. These functions should be fixed in Velox.
* array_min/array_max: https://github.com/facebookincubator/velox/issues/10543.
* array_position: https://github.com/facebookincubator/velox/issues/10580
